### PR TITLE
Fix for incorrect delimiters in relative_input_file on Windows

### PR DESCRIPTION
**Pull request by MichaelOrlov**
- Addressing CI failure described in https://github.com/ros2/rosbag2/pull/1198#issuecomment-1560362070
- Fix for ""// with input from rosbag2_storage_mcap_testdata/msg\\ComplexMsgDependsOnIdl.msg"
- On Windows platform message generator messing up with the input file name ""/msg\\ComplexMsgDependsOnIdl.msg" by inserting double backslashes instead of one forward slash.
- Partial backport from https://github.com/ros2/rosidl/pull/576

### CODE REVIEW

**Review by emersonknapp — COMMENTED** [review]
This looks good to me... but I wonder if it should be considered a breaking change for any existing code depending on the contents of these IDL files within the release

**Comment by MichaelOrlov** [comment]
@emersonknapp This is a good question. I've tried to search in regards changed `relative_input_file` parameter across repos and files where it is used.  
And apparently, it turns out that this parameter only participates in forming a comment message in the generated type description
https://github.com/ros2/rosidl/blob/b62e8c4d156e73f155d3b8aaedaf96462baa94cf/rosidl_adapter/rosidl_adapter/resource/msg.idl.em#L2
i.e.
```
// with input from @(pkg_name)/@(relative_input_file)
```
It will not break any API or ABI compatibility. 
I am very doubt, that anything else depends on the format of the comment `// with input from @(pkg_name)/@(relative_input_file)` in the type description.

**Comment by MichaelOrlov** [comment]
Gist: https://gist.githubusercontent.com/MichaelOrlov/c5c8c547760e183e8f3fb587ac70472f/raw/afd26accee89750545fe9e56bc8256e161a4412b/ros2.repos
BUILD args:  --packages-above-and-dependencies rosidl_adapter rosbag2_storage_mcap
TEST args:  --packages-above rosidl_adapter rosbag2_storage_mcap
ROS Distro: foxy
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/12114
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18798/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13330/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19542/)

**Review by emersonknapp — APPROVED** [review]

**Comment by MichaelOrlov** [comment]
Re-start CI with focal target
Gist: https://gist.githubusercontent.com/MichaelOrlov/c5c8c547760e183e8f3fb587ac70472f/raw/afd26accee89750545fe9e56bc8256e161a4412b/ros2.repos
BUILD args:  --packages-above-and-dependencies rosidl_adapter rosbag2_storage_mcap
TEST args:  --packages-above rosidl_adapter rosbag2_storage_mcap
ROS Distro: foxy
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/12115
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18799/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13331/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19543/)

**Comment by MichaelOrlov** [comment]
Windows build failed due to the issue with the Connext_DDS which is exists in baseline and failing every other Windows job with error messages:
```
Copyright (C) Microsoft Corporation. All rights reserved.



  Checking Build System

  Building Custom Rule C:/ci/ws/src/ros2/rosidl_typesupport_connext/rosidl_typesupport_connext_cpp/CMakeLists.txt

  identifier.cpp

  wstring_conversion.cpp

     Creating library C:/ci/ws/build/rosidl_typesupport_connext_cpp/Release/rosidl_typesupport_connext_cpp.lib and object C:/ci/ws/build/rosidl_typesupport_connext_cpp/Release/rosidl_typesupport_connext_cpp.exp

CVTRES : fatal error CVT1107: 'C:\Program Files\rti_connext_dds-5.3.1\lib\x64Win64VS2017\nddscpp.lib' is corrupt [C:\ci\ws\build\rosidl_typesupport_connext_cpp\rosidl_typesupport_connext_cpp.vcxproj]

LINK : fatal error LNK1123: failure during conversion to COFF: file invalid or corrupt [C:\ci\ws\build\rosidl_typesupport_connext_cpp\rosidl_typesupport_connext_cpp.vcxproj]

---
--- stderr: rosidl_typesupport_connext_cpp
CMake Warning (dev) at C:/ci/ws/install/share/connext_cmake_module/cmake/Modules/FindConnext.cmake:163 (list):
  Policy CMP0121 is not set: The list() command now validates parsing of
  index arguments.  Run "cmake --help-policy CMP0121" for policy details.
  Use the cmake_policy command to set the policy and suppress this warning.
  Invalid list index "nddsc.lib".
Call Stack (most recent call first):
  C:/ci/ws/install/share/connext_cmake_module/cmake/Modules/FindConnext.cmake:198 (_count_found_libraries)
  CMakeLists.txt:25 (find_package)
```

Re-running Windows job one more time without `CI_USE_CONNEXTDDS` flag.
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19546/)

**Comment by MichaelOrlov** [comment]
Windows CI job fails in ` rqt_py_common` with error messages 
```
Traceback (most recent call last):
  File "C:\ci\ws\install\Lib\site-packages\ament_index_python\packages.py", line 50, in get_package_prefix
    content, package_prefix = get_resource('packages', package_name)
  File "C:\ci\ws\install\Lib\site-packages\ament_index_python\resources.py", line 48, in get_resource
    raise LookupError(
LookupError: Could not find the resource 'launch' of type 'packages'
```
I've made some investigation and was able to reproduce this failure on my local machine **with and without** fixes from this PR. It means that this failure exists on the baseline even without my changes. 
One interesting moment is that those tests from `rqt_py_common` fail only if sourcing `local_setup.bat`.
If make a clean build or reopen the console and run
```
colcon test --packages-select rqt_py_common --event-handlers console_direct+
```
The tests will pass without any errors with and without changes from this PR.

**Comment by MichaelOrlov** [comment]
@emersonknapp @clalancette Kindly ping for review and merge. 
I don't have permission to merge into the `rosidl` repo.

**Comment by emersonknapp** [comment]
I don't have that access either - and I would also be hesitant to merge something so low in the stack until CI is all sorted out. I'm running this Windows Foxy build with no changes as a test just to make sure of the baseline [Build Status](https://ci.ros2.org/job/ci_windows/19554/)

**Comment by quarkytale** [comment]
Hey @MichaelOrlov @emersonknapp I can help with this, I'm in for getting things merged but I agree with Emerson, let's have CI sorted out atleast before merging this one.

**Comment by emersonknapp** [comment]
Adding my opinion here - I don't think we should move forward with this change. There will only be one last release, and we don't want to change anything about code generation without having any future chance to fix unexpected consequences. Since this change was intended to fix a test case in a downstream package, I think we should just patch that test to go green for the current behavior, which while maybe odd should be considered correct and final for Foxy

**Comment by quarkytale** [comment]
CI without mcap
[Build Status](https://ci.ros2.org/job/ci_windows/19562/)

**Comment by MichaelOrlov** [comment]
@quarkytale CI still magically fails in `rqt_py_common` We can't merge this PR if it causes regression.
- I've prepared a workaround for the originally failing test from `rosbag2_storage_mcap` in https://github.com/ros2/rosbag2/pull/1355 will see how it will pass CI.

**Comment by MichaelOrlov** [comment]
Ok folks. 
- Windows CI job with https://github.com/ros2/rosbag2/pull/1355 failed the same `rqt_py_common` test with exactly the same error messages:
```
The test did not generate a result file:

Traceback (most recent call last):
  File "C:\ci\ws\install\Lib\site-packages\ament_index_python\packages.py", line 50, in get_package_prefix
    content, package_prefix = get_resource('packages', package_name)
  File "C:\ci\ws\install\Lib\site-packages\ament_index_python\resources.py", line 48, in get_resource
    raise LookupError(
LookupError: Could not find the resource 'launch' of type 'packages'

During handling of the above exception, another exception occurred:
```
- It means that failure in the `rqt_py_common` tests exists on the baseline and that **there is no regression** in the current https://github.com/ros2/rosidl/pull/745 PR. It just turns out that failure in the `rqt_py_common` reproduces when `rosbag2_storage_mcap` also participates in build and test. 
This explains why the test was passing when @emersonknapp was running  CI for pure `master` branch with `--packages-above rosidl_adapter` only without `rosbag2_storage_mcap`.

- Given the fact that with https://github.com/ros2/rosbag2/pull/1355 test failure in the `rqt_py_common` tests will still persist when compiling and running with `rosbag2_storage_mcap` I would strongly recommend to proceed in either of two options.
1. Proceed and merge this https://github.com/ros2/rosidl/pull/745 as an adequate solution for the failing test in `rosbag2_storage_mcap` test. **Please note that it was proved that there is no regression in https://github.com/ros2/rosidl/pull/745**
2. Go ahead with https://github.com/ros2/rosbag2/pull/1347 by reverting mcap bring up entirely and dismiss https://github.com/ros2/rosidl/pull/745 and https://github.com/ros2/rosbag2/pull/1355

**Comment by MichaelOrlov** [comment]
@emersonknapp @quarkytale @clalancette Please see my comment ^^^^ (above).

**Comment by emersonknapp** [comment]
I don't understand why not option 3) Just merge https://github.com/ros2/rosbag2/pull/1355

**Comment by MichaelOrlov** [comment]
> I don't understand why not option 3) Just merge [ros2/rosbag2#1355](https://github.com/ros2/rosbag2/pull/1355)

Because failure in `rqt_py_common` tests will still persist after merging [ros2/rosbag2#1355](https://github.com/ros2/rosbag2/pull/1355). And this is not really a fix for the root cause of the issue, but rather an accommodation of the test to the wrong behavior.

**Comment by emersonknapp** [comment]
I still don't get it - this change #745 doesn't fix `rqt_py_common` either, does it? Are you invested in fixing that `rqt_py_common` test for some reason?

**Comment by MichaelOrlov** [comment]
@emersonknapp Yes. failure in the `rqt_py_common` is totally unrelated to the rosbag2 stuff.

The reason why I prefer #745 versus https://github.com/ros2/rosbag2/pull/1355 is that #745 is really fixing the root cause of the failure in the `rosbag2_storage_mcap` tests and https://github.com/ros2/rosbag2/pull/1355 only making adoptaions in the test for the buggy behaviour.

**Comment by emersonknapp** [comment]
I'm trying to argue that Foxy's codegen should be considered finalized, and since it doesn't affect behavior we can have a little kludge workaround downstream so that the test is not "overfit" to exact text output but instead just tests what matters, which is that things work correctly.

**Comment by MichaelOrlov** [comment]
As regards to the
> Are you invested in fixing that rqt_py_common test for some reason?
I've spent some time trying to figure out why it fails and found that it fails only if sourcing `/install/local_setup.bash` 
I don't know the details of why.

**Comment by MichaelOrlov** [comment]
Why consider Foxy's codegen as finalized if we know that there is obviously a bug and we have an adequate fix for it?

**Comment by quarkytale** [comment]
> ... we have an adequate fix for it?

If that fix actually passes CI and makes the final snapshot stable then that's all I care about.

**Comment by MichaelOrlov** [comment]
> > ... we have an adequate fix for it?
> 
> If that fix actually passes CI and makes the final snapshot stable then that's all I care about.

Though, it can't magically fix what was already broken before in the `rqt_py_common`.
The #745 intends to fix failures in `rosbag2_storage_mcap` tests and actually fixes them.

**Comment by MichaelOrlov** [comment]
- Closing as will not be merged in favor of the https://github.com/ros2/rosbag2/pull/1347
- We decided to revert mcap bring up completely to avoid any risks and concerns.

**Pull request closed without merging**
